File: python/Sliding Window/longest_repeating_character_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        """
        Find the length of the longest substring that can be obtained by replacing at most k characters so all characters are the same.
        Approach:
        - Use sliding window and a hashmap to count character frequencies.
        - Track the max frequency in the window.
        - If window size minus max frequency > k, shrink window from the left.
        Time Complexity: O(n)
        Space Complexity: O(1) (since alphabet size is constant)
        """
        # Dictionary to store count of each character in current window
        count = {}
        res = 0
        l = 0
        max_freq = 0  # Track maximum frequency in current window

        for r in range(len(s)):
            # Update count of current character
            count[s[r]] = 1 + count.get(s[r], 0)
            # Update max frequency
            max_freq = max(max_freq, count[s[r]])

            # If window size minus max frequency is greater than k,
            # we need to shrink the window
            while (r - l + 1) - max_freq > k:
                count[s[l]] -= 1
                l += 1

            # Update result with current window size
            res = max(res, r - l + 1)
        
        return res

# -------------------
# Test Cases
# -------------------
def test_character_replacement():
    s = Solution()
    # Test case 1: Basic
    assert s.characterReplacement("ABAB", 2) == 4
    # Test case 2: Replace all
    assert s.characterReplacement("AABABBA", 1) == 4
    # Test case 3: No replacements needed
    assert s.characterReplacement("AAAA", 2) == 4
    # Test case 4: k = 0
    assert s.characterReplacement("AABABBA", 0) == 2
    # Test case 5: Empty string
    assert s.characterReplacement("", 2) == 0
    # Test case 6: All unique
    assert s.characterReplacement("ABCDE", 1) == 2
    print("All test cases passed!")

if __name__ == "__main__":
    test_character_replacement()

File: C++/Sliding Window/longest_repeating_character_replacement.cpp
#include <bits/stdc++.h>        
using namespace std;

class Solution {
public:
    /*
     * Problem: Find longest substring where k character replacements make all chars same
     * 
     * Approach:
     * - Use sliding window with character frequency map
     * - Track maximum frequency in current window
     * - If (window size - max frequency) > k, shrink window
     * - Update result with valid window sizes
     * 
     * Time: O(n) - single pass through string
     * Space: O(1) - array of size 26 for uppercase letters
     */
    int characterReplacement(string s, int k) {
        // Array to count frequencies (for uppercase letters A-Z)
        vector<int> count(26, 0);
        int result = 0;
        int left = 0;
        int maxFreq = 0;
        
        // Expand window with right pointer
        for (int right = 0; right < s.length(); right++) {
            // Update frequency of current character
            count[s[right] - 'A']++;
            // Update maximum frequency in current window
            maxFreq = max(maxFreq, count[s[right] - 'A']);
            
            // If window invalid (too many replacements needed)
            while ((right - left + 1) - maxFreq > k) {
                // Decrease frequency of leftmost char
                count[s[left] - 'A']--;
                left++;
            }
            
            // Update result with current valid window size
            result = max(result, right - left + 1);
        }
        
        return result;
    }
};

/*
 * Detailed C++ Syntax Explanation for Beginners:
 * 
 * 1. Vector Initialization:
 *    vector<int> count(26, 0)  // Size 26 array filled with zeros
 *    - Used instead of Python dictionary
 *    - Index 0-25 represents A-Z
 * 
 * 2. Character to Index:
 *    s[right] - 'A'    // Convert char to 0-25 index
 *    - 'A' is ASCII 65, so 'A'-'A'=0, 'B'-'A'=1, etc.
 * 
 * 3. String Operations:
 *    s.length()        // Get string length
 *    s[index]          // Access character at position
 * 
 * 4. Window Size:
 *    right - left + 1  // Current window size
 *    - +1 because positions are zero-based
 * 
 * 5. Standard Library:
 *    max(a, b)         // Returns larger of two values
 * 
 * 6. Variables:
 *    maxFreq           // Most frequent char count
 *    result           // Longest valid substring
 * 
 * Note: This implementation:
 * - Uses array instead of hashmap for O(1) space
 * - Maintains O(n) time complexity
 * - Handles uppercase English letters only
 */
